[Feat] Prometheus metric export

sharonsyh opened the pull request:
This pull request introduces Prometheus-based metric tracking for energy and power usage within the Zeus framework. It includes functionality for monitoring GPU, CPU, and DRAM energy usage via Histograms, Cumulative Counters, and Gauges.

- **`zeus/metric.py`**:  
  A new module that introduces `EnergyHistogram`, `EnergyCumulativeCounter`, and `PowerGauge` classes. These classes enable real-time monitoring of CPU, GPU, and DRAM energy and power consumption by integrating with Prometheus.

- **`zeus/prometheus.yml`**:  
  Configuration file for setting up Prometheus monitoring.

- **`zeus/docker-compose.yml`**:  
  A Docker Compose file for easily setting up Prometheus with the project for local or cloud-based monitoring.

- **Modified `pyproject.toml`**:  
  Added `prometheus-client` as an optional dependency for Prometheus metric integration.

sharonsyh committed:
Add metric.py, prometheus configs, and modify pyproject.toml

sharonsyh committed:
Reformat metric.py with black

sharonsyh committed:
Add metric monitoring section to documentation

sharonsyh committed:
Add unit tests for EnergyHistogram, EnergyCumulativeCounter, and PowerGauge

sharonsyh committed:
Add train_single.py for testing energy monitoring metrics

jaywonchung commented on docker/prometheus/prometheus.yml:
```diff
@@ -0,0 +1,14 @@
+      - targets: ['localhost:9090']  
+  - job_name: 'pushgateway'
+    static_configs:
+      - targets: ['zeus-pushgateway-1:9091']
```
What is `zeus-pushgateway-1`? Is this something the users should configure? I felt like from the docker compose file, it should be called just `pushgateway`.

jaywonchung commented on docs/measure/index.md:
````diff
@@ -114,6 +114,88 @@ if __name__ == "__main__":
+
+You can customize the bucket ranges for each component (GPU, CPU, and DRAM), or let Zeus use default ranges.
+
+```python hl_lines="2 5-7"
````
```suggestion
```python hl_lines="2 5-15"
```
You can preview documentation with `bash scripts/preview_docs.sh`. The script will also report missing docstrings and stuff.

jaywonchung commented on docs/measure/index.md:
```diff
@@ -114,6 +114,88 @@ if __name__ == "__main__":
+if __name__ == "__main__":
+    # Initialize EnergyHistogram with custom bucket ranges
+    histogram_metric = EnergyHistogram(
+        energy_monitor=ZeusMonitor,
```
This should be an actual instance of a ZeusMonitor.

jaywonchung commented on docs/measure/index.md:
```diff
@@ -114,6 +114,88 @@ if __name__ == "__main__":
+
+    histogram_metric.begin_window("histogram_test")
+    # Perform tasks
+    histogram_metric.end_window("histogram_test")
```
Give it a more concrete example. I would write a simplified training loop in PyTorch (see above in this file for reference) and then have the histogram metric collect a histogram of the energy consumption of one training epoch.

jaywonchung commented on examples/prometheus/train_single.py:
```diff
@@ -0,0 +1,409 @@
+import torchvision.models as models
+from zeus.metric import EnergyHistogram
+from zeus.metric import EnergyCumulativeCounter
+from zeus.metric import PowerGauge
```
Merge and move to line 29.

jaywonchung commented on examples/prometheus/train_single.py:
```diff
@@ -0,0 +1,409 @@
+    # ZeusMonitor is used to profile the time and energy consumption of the GPU.
+    energy_monitor = ZeusMonitor(gpu_indices=[args.gpu])
+    power_monitor = PowerMonitor(gpu_indices=[args.gpu])
+    energy_histogram = EnergyHistogram(energy_monitor=energy_monitor, prometheus_url='http://localhost:9091', job='training_energy_histogram')
```
Have a concise comment explaining what it is before each metric.

jaywonchung commented on examples/prometheus/train_single.py:
```diff
@@ -0,0 +1,409 @@
+    power_gauge.begin_window("Gauge Window")
+
+    energy_counter.end_window("Counter_window")
+    power_gauge.end_window("Gauge Window")
```
I don't get this part. Why are you starting a window and then immediately ending it?

jaywonchung replied:
1. Please use more descriptive names for metrics. Like epoch energy.
2. Is it better to have underscores in metric names? Or human-readable names? Or is either fine?

sharonsyh replied:
> I don't get this part. Why are you starting a window and then immediately ending it?

The current begin_window and end_window calls were placeholders for testing due to a multiprocessing-related error. Once I resolve the issue, I’ll replace this with meaningful usage.

sharonsyh replied:
> 1. Please use more descriptive names for metrics. Like epoch energy.
> 2. Is it better to have underscores in metric names? Or human-readable names? Or is either fine?

For the metric names, it is said on the documentation that Prometheus metrics and label names are written in snake_case and converting camelCase to snake_case is desirable!

jaywonchung replied:
Let's let the users decide their metric/label names and not do any extra formatting inside Zeus. But, in the examples we provide, we should follow best practices -- underscored names.

jaywonchung commented on examples/prometheus/train_single.py:
```diff
@@ -0,0 +1,409 @@
+
+
+if __name__ == "__main__":
+    main()
```
EOL missing.

jaywonchung commented on tests/test_metric.py:
```diff
@@ -0,0 +1,204 @@
+from unittest.mock import patch
+
+import pytest
+from zeus.metric import EnergyHistogram, EnergyCumulativeCounter, PowerGauge
```
```suggestion
from __future__ import annotations

from unittest.mock import patch

import pytest

from zeus.metric import EnergyHistogram, EnergyCumulativeCounter, PowerGauge
```

Imports should have four clusters separated by a newline:
1. Imports from `__future__`
2. Imports from built-in modules
3. Imports from third party dependencies
4. Imports from myself (i.e., Zeus)

jaywonchung commented on tests/test_metric.py:
```diff
@@ -0,0 +1,204 @@
+    Mock object representing energy measurements for testing.
+    Contains energy values for GPU, CPU, and DRAM.
+    """
+    def __init__(self, gpu_energy=None, cpu_energy=None, dram_energy=None):
```
```suggestion
    def __init__(
        self,
        gpu_energy: dict | None = None,
        cpu_energy: dict | None = None,
        dram_energy: dict | None = None,
    ) -> None:
```

All Zeus code should be typed.

jaywonchung commented on tests/test_metric.py:
```diff
@@ -0,0 +1,204 @@
+    """
+    Mock object representing energy measurements for testing.
+    Contains energy values for GPU, CPU, and DRAM.
+    """
```
You're using a docstring style that is inconsistent with the rest of the codebase. Please skim the docstring section of Google's Python style guide [here](https://android.googlesource.com/platform/external/google-styleguide/+/refs/tags/android-s-beta-2/pyguide.md#3_8-comments-and-docstrings).

jaywonchung commented on tests/test_metric.py:
```diff
@@ -0,0 +1,204 @@
+        return MockMeasurement(
+            gpu_energy={index: 30.0 for index in self.gpu_indices},  # Fixed value for all GPUs
+            cpu_energy={index: 15.0 for index in self.cpu_indices},  # Fixed value for all CPUs
+            dram_energy={index: 7.5 for index in self.dram_indices}  # Fixed value for all DRAMs
```
This part. Also lines 24 and 25. All the comments are more or less meaningless.

Comments are only useful for non-trivial stuff where things aren't obvious from the source code itself. If you write comments for everything, it just increases complexity for the person reading the code. Please only comment things that you think are non-trivial for a first-time reader. (An exception for this is docstrings for functions/classes/modules. They should always exist for documentation purposes.)

jaywonchung commented on zeus/metric.py:
```diff
@@ -0,0 +1,477 @@
+    Counter,
+    Gauge,
+    push_to_gateway,
+)
```
Import formatting needed (see previous comment on import clustering)

jaywonchung commented on zeus/metric.py:
```diff
@@ -0,0 +1,477 @@
+)
+
+
+class Metric(abc.ABC):
```
Docstring missing.

jaywonchung commented on zeus/metric.py:
```diff
@@ -0,0 +1,477 @@
+
+
+class EnergyHistogram(Metric):
+    """EnergyHistogram class to monitor and record energy consumption metrics.
```
```suggestion
    """Measures the energy consumption a code range and exports a histogram metric.
```

jaywonchung commented on docs/measure/index.md:
```diff
@@ -114,6 +114,88 @@ if __name__ == "__main__":
+            "gpu": [10.0, 25.0, 50.0, 100.0],
+            "cpu": [5.0, 10.0, 25.0, 50.0],
+            "dram": [1.0, 2.5, 5.0, 10.0]
+        }
```
I think we can remove this from the example for simplicity. We also have a separate `!!! tip` box for this argument.

jaywonchung commented on zeus/metric.py:
```diff
@@ -0,0 +1,477 @@
+        energy_monitor: ZeusMonitor,
+        prometheus_url: str,
+        job: str,
+        bucket_ranges=None,
```
This is not type safe. In other words, if the user makes a mistake like this

```python
bucket_ranges={
    "tpu": [10.0, 25.0, 50.0, 100.0],
}
```

nothing protects the user from making a mistake this (and typos are common in programming). Code will silently run (without erroring out) with the default bucket range and the user will have a miserable time debugging.

While it is repetitive and somewhat not very clean, I think the best solution is to just have

```suggestion
        gpu_bucket_range: list[float] = [50.0, 100.0, 200.0, 500.0, 1000.0],
        cpu_bucket_range: list[float] = [10.0, 20.0, 50.0, 100.0, 200.0],
        dram_bucket_range: list[float] = [5.0, 10.0, 20.0, 50.0, 150.0],
```

This has two advantages:

1. It is type-safe. It is impossible for the user to make the mistake that I illustrated earlier.
2. The default bucket range is evident from the signature of this function.

jaywonchung commented on zeus/metric.py:
```diff
@@ -0,0 +1,477 @@
+                )
+            # Only when CPUs are available, we check if DRAM is available using supportsGetDramEnergyConsumption in CPU class
+            # If DRAM available, we create histogram for each DRAM indices for each CPU indices
+            if CPU.supportsGetDramEnergyConsumption:
```
Seems to be a typo. `supportsGetDramEnergyConsumption` is a method on `CPU`.

You should check this for each cpu index. Reference:

```python
for i, cpu in enumerate(get_cpus()):
    if cpu.supportsGetDramEnergyConsumption():
        print(f"DRAM energy measurement is supported on CPU {i}").
```

jaywonchung commented on zeus/metric.py:
```diff
@@ -0,0 +1,477 @@
+                self.gpu_histograms[gpu_index] = Histogram(
+                    f"energy_monitor_gpu_{gpu_index}_energy_joules",
+                    f"GPU {gpu_index} energy consumption",
+                    ["window", "index"],
```
Question: What does this mean?

sharonsyh replied:
The first line, `f"energy_monitor_gpu_{gpu_index}_energy_joules"`, defines the metric name according to Prometheus naming conventions.

The second line, `f"GPU {gpu_index} energy consumption"`, is the description of what the metric measures, 

The third line, `["window", "index"]`, specifies the labels for this metric:
- window: Identifies the phase or range of the monitoring period (e.g., window="epoch_0_energy" represents the energy consumption during the 0th epoch).
- index: Refers to the specific GPU module being monitored. This label helps differentiate metrics when monitoring multiple GPUs, allowing precise identification and querying for individual components.

jaywonchung commented on zeus/metric.py:
```diff
@@ -0,0 +1,477 @@
+                    f"energy_monitor_gpu_{gpu_index}_energy_joules",
+                    f"GPU {gpu_index} energy consumption",
+                    ["window", "index"],
+                    buckets=self.bucket_ranges.get("gpu", []),
```
What happens when `buckets` ends up being `[]`?

jaywonchung commented on zeus/metric.py:
```diff
@@ -0,0 +1,477 @@
+        """Begin a new energy monitoring window."""
+        self.energy_monitor.begin_window(f"__EnergyHistogram_{name}")
+
+    def end_window(self, name: str) -> None:
```
In all of the metrics, please mention somewhere that the name of the window **must** match between calls to `begin_window` and `end_window`.

jaywonchung commented on zeus/metric.py:
```diff
@@ -0,0 +1,477 @@
+            for cpu_index in self.energy_monitor.cpu_indices:
+                self.cpu_histograms[cpu_index] = Histogram(
+                    f"energy_monitor_cpu_{cpu_index}_energy_joules",
+                    f"CPU {cpu_index} energy consumption",
```
Question: Are these the name of the metric and its description?

If so, I think these should also be documented. That is, given cpu indices and window names, how the exact name of the metrics from our metric classes are constructed.

Also, do you think the user might want to override this based on their needs? I think I currently have a poor understanding of how metric names are organized in Prometheus. How are these two parameters, `job`, and the window name organized from the user's perspective, looking at their Prometheus dashboard?

All these ought to be documented. Instead of answering this comment, you can just update the documentation and let me know.

jaywonchung commented on zeus/metric.py:
```diff
@@ -0,0 +1,477 @@
+                if gpu_energy > self.max_gpu_bucket:
+                    warnings.warn(
+                        f"GPU {gpu_index} energy {gpu_energy} exceeds the maximum bucket value of {self.max_gpu_bucket}"
+                    )
```
Very nice

jaywonchung commented on zeus/metric.py:
```diff
@@ -0,0 +1,477 @@
+            for cpu_index, cpu_energy in measurement.cpu_energy.items():
+                if cpu_index in self.cpu_histograms:
+                    self.cpu_histograms[cpu_index].labels(
+                        window=f"__EnergyHistogram_{name}", index=cpu_index
```
```suggestion
                        window=name, index=cpu_index
```
I'm assuming `window` is a label that is displayed publicly to the user. `__EnergyHistogram_` is what we internally prepended to the window name to make it unique w.r.t. the energy monitor, so it should not be exposed. What the user passed into `end_window` is the name of the window the user wants.

jaywonchung commented on zeus/metric.py:
```diff
@@ -0,0 +1,477 @@
+            name: A unique name for the monitoring window (used as a label in Prometheus metrics).
+        """
+        self.queue = mp.Queue()
+        self.proc = mp.Process(
```
Explore if using the following prevents the error.

```suggestion
        context = mp.get_context("spawn")
        self.queue = context.Queue()
        self.proc = context.Process(
```

Please make sure you're understanding what this is doing.

jaywonchung commented on zeus/metric.py:
```diff
@@ -0,0 +1,477 @@
+                registry=registry,
+            )
+
+        if CPU.supportsGetDramEnergyConsumption:
```
ditto

jaywonchung reviewed (commented):
Thanks for the great work! This is an important piece in making Zeus more usable in a real world scenario. I looked over it at a mid- to high-level (not the nitty gritty details yet) and left some comments. Let me know what you think.

sharonsyh committed:
Update docs/measure/index.md

Co-authored-by: Jae-Won Chung <jwnchung@umich.edu>

jaywonchung commented on examples/prometheus/requirements.txt:
Please create a README.md similar to other examples.

jaywonchung commented on docs/measure/index.md:
````diff
@@ -114,6 +114,88 @@ if __name__ == "__main__":
         avg_energy = sum(map(lambda m: m.total_energy, steps)) / len(steps)
         print(f"One step takes {avg_time} s and {avg_energy} J for the CPU.")
 ```
+## Metric Monitoring
````
We need to have a section on our assumptions on Prometheus. For instance, we assume that the Prometheus Push Gateway is deployed, so we need to talk about that and have a brief section for how to do that or a link to an official documentation. Are there other assumptions?

sharonsyh committed:
Refactor metric initialization and multiprocessing logic in metric.py

- Changed metric instantiation to accept CPU and GPU indices directly instead of class objects.
- Improved multiprocessing logic to address and fix pickle-related errors.
- Added consistent handling for sync_execution across begin_window and end_window calls for all metrics.
- Centralized bucket range validation and default handling for EnergyHistogram.
- Improved error handling and logging for multiprocessing processes.
- Standardized Prometheus metric labels (e.g., window and index) across Histogram, Counter, and Gauge.
- Updated docstrings for consistency and clarity across all Metric subclasses.

sharonsyh committed:
Update prometheus.yml

Adjust target names to standardize pushgateway references, ensuring consistency with the Docker Compose configuration.

sharonsyh committed:
Improve example training script to include Zeus metrics

sharonsyh committed:
Remove unintended file tests/test_metric.py from repository

sharonsyh committed:
Update the doc on Metrics Monitoring and Assumptions

sharonsyh committed:
Update index.md

sharonsyh committed:
Update index.md

sharonsyh committed:
Add README for example training file with Zeus energy metrics integration

sharonsyh committed:
Add Metric Name Construction section on index.md

sharonsyh committed:
Update index.md

sharonsyh committed:
Update README.md to include the dependency on prometheus_client

sharonsyh committed:
Update unit tests for the modified metric.py

sharonsyh committed:
Fix formatting issues detected by black

sharonsyh committed:
Fix formatting issues detected by black

sharonsyh committed:
Fix formatting issues detected by black

sharonsyh committed:
Fix formatting issues detected by black

sharonsyh committed:
Resolve unbound variable errors

sharonsyh committed:
Resolve unbound variable errors

sharonsyh committed:
Specify type for the args

sharonsyh committed:
Fix formatting issues detected by black

jaywonchung committed:
Merge master

sharonsyh committed:
Fix energy histogram to properly handle default bucket ranges

sharonsyh committed:
Add the mock_push_to_gateway Patch to each test

sharonsyh committed:
Update gpu_bucket_range, cpu_bucket_range, and dram_bucket_range in the EnergyHistogram constructor to use Sequence[float] for type safety, maintaining default values for bucket ranges

sharonsyh committed:
Patch to mock urllib.request.urlopen preventing attempts to an actual HTTP request

sharonsyh committed:
Patch to mock urllib.request.urlopen preventing attempts to an actual HTTP request

sharonsyh committed:
Patch to mock prometheus_client.exposition.push_to_gateway external calls

sharonsyh committed:
Patch to http.client.HTTPConnection

sharonsyh committed:
Remove unneccessary mock

jaywonchung commented on zeus/metric.py:
```diff
@@ -0,0 +1,532 @@
+    """
+
+    @abc.abstractmethod
+    def begin_window(self, name: str):
```
Both `begin_window` and `end_window` must accept `sync_execution: bool = True`, following `ZeusMonitor`. Then, `sync_execution` should be passed in as is to `ZeusMonitor.begin_window` or `ZeusMonitor.end_window`.

jaywonchung commented on docs/measure/index.md:
```diff
@@ -114,6 +114,236 @@ if __name__ == "__main__":
+
+    for epoch in range(100):
+        # Start monitoring energy for the entire epoch
+        energy_histogram.begin_window(f"epoch_{epoch}_energy")
```
This got me confused. Why is it okay for the window name to be different on each epoch? I thought the window name should always be the same (e.g., `training_epoch`), and each time when a window with exactly that name is measured (in this example snippet, every epoch), it will add a data point to the histogram metric uniquely identified by that window name.

jaywonchung replied:
This is the largest concern of this review round. I see below that you also have:

```python
energy_histogram.begin_window(f"epoch_{epoch}_step_{step_idx}_energy")
```

And you're using the same instance of `energy_histogram`. I hope I'm wrong, but this has to mean that you're bunching together energy measurements from both training epoch and training step into the same `Histogram` metric with just different `window` labels, aren't you? If so, this defeats the whole purpose of histogram monitoring. The result of this example should be two completely independent `Histogram` metrics, one for monitoring training epochs (one datapoint added whenever `end_window("training_epoch")` is called) and another for monitoring training steps (one datapoint added whenever `end_window("training_step")` is called).

jaywonchung commented on examples/prometheus/README.md:
````diff
@@ -0,0 +1,55 @@
+    ```
+
+## EnergyHistogram, PowerGauge, and EnergyCumulativeCounter
+- [`EnergyHistogram`][zeus.metric.EnergyHistogram]: Records energy consumption data for GPUs, CPUs, and DRAM and pushes the data to Prometheus as histogram metrics. This is useful for tracking energy usage distribution over time.
````
If you check how Markdown is rendered on GitHub ([here](https://github.com/ml-energy/zeus/blob/f032488d9c1d98375db44fa7a919eda79a064a00/examples/prometheus/README.md)), you'll see that this kind of link is not rendered correctly. The bracket syntax is only for Mkdocs. GitHub isn't smart enough to automatically understand what `zeus.metric.EnergyHistogram` is and link to the related reference doc. Instead, you should manually link the corresponding API documentation link:

```markdown
[`ZeusMonitor`](https://ml.energy/zeus/reference/monitor/energy/#zeus.monitor.energy.ZeusMonitor)
```

Since `zeus.metrics` has not landed in `master` yet, there is no corresponding live API reference documentation under https://ml.energy/zeus. Thus you should open up `scripts/preview_docs.sh` and construct the right URL for the three classes.

jaywonchung commented on examples/prometheus/README.md:
````diff
@@ -0,0 +1,55 @@
+1. Install `prometheus_client`:
+    ```sh
+    pip install prometheus-client
+    ```
````
Update all installation instructions to `pip install zeus-ml[prometheus]`.

jaywonchung commented on examples/prometheus/train_single.py:
```diff
@@ -0,0 +1,432 @@
+
+    end = time.time()
+    for i, (images, target) in enumerate(train_loader):
+        power_limit_optimizer.on_step_begin()  # Mark the beginning of one training step.
```
All instances of `power_limit_optimizer` should be removed. You probably forgot to remove it after copying this from another directory.

jaywonchung commented on examples/prometheus/train_single.py:
```diff
@@ -0,0 +1,432 @@
+
+def train(
+    train_loader, model, criterion, optimizer, epoch, args, power_limit_optimizer
+):
```
Wait where is `train` called?? How did this example work without doing any training?

jaywonchung commented on pyproject.toml:
```diff
@@ -41,15 +42,16 @@ Documentation = "https://ml.energy/zeus"
-test = ["fastapi[all]", "sqlalchemy", "pydantic<2", "pytest==7.3.2", "pytest-mock==3.10.0", "pytest-xdist==3.3.1", "anyio==3.7.1", "aiosqlite==0.20.0", "numpy<2"]
+test = ["fastapi[standard]", "sqlalchemy", "pydantic<2", "pytest==7.3.2", "pytest-mock==3.10.0", "pytest-xdist==3.3.1", "anyio==3.7.1", "aiosqlite==0.20.0", "numpy<2"]
 docs = ["mkdocs-material[imaging]==9.5.19", "mkdocstrings[python]==0.25.0", "mkdocs-gen-files==0.5.0", "mkdocs-literate-nav==0.6.1", "mkdocs-section-index==0.3.9", "mkdocs-redirects==1.2.1", "urllib3<2", "black"]
+prometheus = ["prometheus-client"]
```
Move this line up to right below `migration`.

jaywonchung commented on zeus/metric.py:
```diff
@@ -0,0 +1,532 @@
+import warnings
+import multiprocessing as mp
+
+from typing import Sequence
```
```suggestion
import multiprocessing as mp
from typing import Sequence
```

jaywonchung commented on zeus/metric.py:
```diff
@@ -0,0 +1,532 @@
+from zeus.monitor.power import PowerMonitor
+from zeus.monitor.energy import ZeusMonitor
+
+from zeus.device.cpu import get_cpus
```
```suggestion
from zeus.monitor.energy import ZeusMonitor
from zeus.device.cpu import get_cpus
```

jaywonchung commented on zeus/metric.py:
```diff
@@ -0,0 +1,532 @@
+                    self.gpu_histograms[gpu_index].labels(
+                        window=name, index=gpu_index
+                    ).observe(gpu_energy)
+                if gpu_energy > self.max_gpu_bucket:
+                    warnings.warn(
+                        f"GPU {gpu_index} energy {gpu_energy} exceeds the maximum bucket value of {self.max_gpu_bucket}",
+                        stacklevel=1,
+                    )
```
There should be a warning when the measured energy is smaller than the minimum bucket as well.

jaywonchung commented on zeus/metric.py:
```diff
@@ -0,0 +1,532 @@
+        """
+        context = mp.get_context("spawn")
+        self.queue = context.Queue()
+        self.proc = context.Process(
```
This will break when users have more than one concurrent monitoring windows, as the second call to `begin_window` will overwrite `self.queue` and `self.proc`. **An arbitrary number of windows can overlap in time in arbitrary ways.** That is why `ZeusMonitor` keeps an internal state for each window with a `dict` key'ed by the name of the window. For this metric, the state of the window is probably the queue and process handles.

This comment also applies to `PowerGauge`.

jaywonchung replied:
Something like this:

```python
@dataclass
class MonitoringProcessState:
    queue: mp.Queue
    proc: mp.Process

# Inside __init__
self.window_state: dict[str, MonitoringProcessState] = {}

# Inside `begin_window`
if window_name in self.window_state:
    raise ValueError(f"Energy cumulative counter metric '{window_name}' already exists")
queue = ...
proc = ...
self.window_state[window_name] = MonitoringProcessState(queue, proc)
```

jaywonchung reviewed (changes requested):
Thanks for the work @sharonsyh! Unfortunately, I think I see design flaws in each of the three metrics. I've left comments, and we should definitely talk about these on Monday.

sharonsyh committed:
Add zeus.metric to the list

sharonsyh committed:
Update reference link for each class

sharonsyh committed:
Move line for prometheus-client

sharonsyh committed:
feat: Add multiprocessing dict and sync execution for begin/end window

- Introduced `MonitoringProcessState` dataclass to handle each multiprocessing process.
- Added `sync_execution` parameter to `begin_window` and `end_window` methods in `Metric` and its subclasses.
- Updated `EnergyHistogram`, `EnergyCumulativeCounter`, and `PowerGauge` to handle `sync_execution` dynamically.
- Modified `energy_monitoring_loop` and `power_monitoring_loop` to include process-safe handling.
- Improved process cleanup and error handling for monitoring windows.

sharonsyh committed:
Add error handling for queue

sharonsyh committed:
Add a call to train() in main

sharonsyh committed:
Refactor tests for the modified code

sharonsyh committed:
Reformat the metric monitoring section for consistency

sharonsyh committed:
Setup Guide -> Local Setup Guide

sharonsyh committed:
Add condition for using put() with empty queue

sharonsyh committed:
Import the SpawnProcess class from multiprocessing.context

jaywonchung commented on docs/measure/index.md:
```diff
@@ -116,55 +116,17 @@ if __name__ == "__main__":
-### **Metric Name Construction**
-Zeus organizes metrics using **static metric names** and **dynamic labels** to ensure flexibility and ease of querying in Prometheus. Below, we document how metric names are constructed and how users can query them effectively.
+!!! Assumption
+    A Prometheus Push Gateway must be deployed and accessible at the URL provided in your Zeus configuration. **This ensures that metrics collected by Zeus can be pushed to Prometheus.**
```
```suggestion
    A [Prometheus Push Gateway](https://prometheus.io/docs/instrumenting/pushing/) must be deployed and accessible. This ensures that metrics collected by Zeus can be pushed to Prometheus.
```

jaywonchung commented on docs/measure/index.md:
```diff
@@ -212,16 +174,25 @@ scrape_configs:
+#### Metric Name
+- For Histogram: `energy_monitor_{component}_energy_joules`
+- For Counter: `energy_monitor_{component}_energy_joules`
+- For Gauge: `power_monitor_gpu_power_watts`
```
```suggestion
- For Gauge: `power_monitor_{component}_power_watts`

Note that Gauge only supports the GPU component at the moment. Tracking issue: [#128](https://github.com/ml-energy/zeus/issues/128)

```

jaywonchung commented on docs/measure/index.md:
````diff
@@ -212,16 +174,25 @@ scrape_configs:
-```python hl_lines="2 5-15"
+#### Labels
+- window: The user-defined window name provided during `begin_window()` and `end_window()` (e.g., `energy_histogram.begin_window(f"epoch_energy")`).
+- index: The GPU index (e.g., `0` for GPU 0).
````
```suggestion
- index: The index of the device (e.g., `0` for GPU 0).
```

jaywonchung commented on docs/measure/index.md:
```diff
@@ -235,46 +206,37 @@ if __name__ == "__main__":
+        job='training_energy_histogram',
+        gpu_bucket_range = [10.0, 25.0, 50.0, 100.0],
+        cpu_bucket_range = [5.0, 15.0, 30.0, 50.0],
+        dram_bucket_range = [2.0, 8.0, 20.0, 40.0],
```
```suggestion
        gpu_bucket_range=[10.0, 25.0, 50.0, 100.0],
        cpu_bucket_range=[5.0, 15.0, 30.0, 50.0],
        dram_bucket_range=[2.0, 8.0, 20.0, 40.0],
```

jaywonchung commented on docs/measure/index.md:
```diff
@@ -235,46 +206,37 @@ if __name__ == "__main__":
+        cpu_indices=[0], 
+        gpu_indices=[0], 
+        prometheus_url='http://localhost:9091', 
+        job='training_energy_histogram',
```
```suggestion
        prometheus_url="http://localhost:9091", 
        job="training_energy_histogram",
```

jaywonchung commented on docs/measure/index.md:
````diff
@@ -235,46 +206,37 @@ if __name__ == "__main__":
+- GPU: [50.0, 100.0, 200.0, 500.0, 1000.0]
+- CPU: [10.0, 20.0, 50.0, 100.0, 200.0]
+- DRAM: [5.0, 10.0, 20.0, 50.0, 150.0]
+```
````
```suggestion

- GPU: `[50.0, 100.0, 200.0, 500.0, 1000.0]`
- CPU: `[10.0, 20.0, 50.0, 100.0, 200.0]`
- DRAM: `[5.0, 10.0, 20.0, 50.0, 150.0]`
```

jaywonchung commented on docs/measure/index.md:
````diff
@@ -345,6 +307,22 @@ if __name__ == "__main__":
+#### Query to Sum Energy for Specific GPU Across All Windows
+```promql
+sum(energy_monitor_gpu_energy_joules{index="0"})
+```
````
````suggestion
Energy for a specific window:
```promql
energy_monitor_gpu_energy_joules{window="epoch_energy"}
```

Sum of energy for a specific window:
```promql
sum(energy_monitor_gpu_energy_joules) by (window)
```

Sum of energy for specific GPU across all windows:
```promql
sum(energy_monitor_gpu_energy_joules{index="0"})
```
````

jaywonchung commented on examples/prometheus/train_single.py:
```diff
@@ -19,6 +19,8 @@
 import torchvision.models as models
 from multiprocessing import set_start_method
+from PIL import Image, ImageFile, UnidentifiedImageError
+#set_start_method("fork", force=True)
```
Please remove the random image-related stuff that were added and completely remove the power limit optimizer, not commenting it out. Overall, this file needs cleanup.

jaywonchung commented on examples/prometheus/train_single.py:
```diff
@@ -180,64 +200,32 @@ def main():
+    # Gauge to track power consumption over time
+    power_gauge = PowerGauge(gpu_indices=[0], update_period=2, prometheus_url='http://localhost:9091', job='training_power_gauge')
+    # Counter to track energy consumption over time
+    energy_counter = EnergyCumulativeCounter(cpu_indices=[0,1], gpu_indices=[0], update_period=2, prometheus_url='http://localhost:9091', job='training_energy_counter')
```
These lines are very long. They're better off using multiple lines, like before the change.

jaywonchung commented on examples/prometheus/train_single.py:
```diff
@@ -180,64 +200,32 @@ def main():
-        
+        energy_histogram.begin_window("training_energy")
+        energy_histogram.end_window("training_energy")
+        train(train_loader, model, criterion, optimizer, epoch, args)
```
Why isn't this `begin_window`, `train`, then `end_window`?

jaywonchung commented on examples/prometheus/train_single.py:
```diff
@@ -180,64 +200,32 @@ def main():
+        train(train_loader, model, criterion, optimizer, epoch, args)
         print(f"Top-1 accuracy: {acc1}")
 
-    # Allow metrics to capture remaining data before shutting down monitoring.
```
These comments are useful. Please bring them back.

jaywonchung commented on examples/prometheus/train_single.py:
```diff
@@ -430,3 +418,4 @@ def accuracy(output, target, topk=(1,)):
 
 if __name__ == "__main__":
     main()
+    
```
Newline.

jaywonchung commented on tests/test_metric.py:
```diff
@@ -25,8 +28,8 @@ def mock_zeus_monitor():
-            cpu_energy={0: 20.0, 1: 25.0},
+            gpu_energy={0: 50.0, 1: 100.0, 2: 200.0},
+            cpu_energy={0: 40.0, 1: 50.0},
             dram_energy={},
```
If mock CPU 0 supports DRAM energy measurement (in `mock_get_cpus`), shouldn't this be something like `dram_energy={0: 10.0}`?

The metrics would be expecting the monitor to provide DRAM energy measurements for CPU 0, but if the Measurement object has nothing, shouldn't it raise an error?

jaywonchung commented on zeus/metric.py:
```diff
@@ -31,13 +40,23 @@ class Metric(abc.ABC):
+
+        Args:
+            name (str): Name of the measurement window.
+            sync_execution (bool): Whether to execute synchronously. Defaults to None.
```
This is wrong. See `ZeusMonitor`.

jaywonchung commented on zeus/metric.py:
```diff
@@ -54,6 +73,9 @@ class EnergyHistogram(Metric):
         dram_bucket_range: Histogram buckets for DRAM energy.
+        gpu_histograms: A single Prometheus Histogram metric for all GPU energy consumption, indexed by window and GPU index.
+        cpu_histograms: A single Prometheus Histogram metric for all CPU energy consumption, indexed by window and CPU index.
+        dram_histograms: A single Prometheus Histogram metric for all DRAM energy consumption, indexed by window and DRAM index.
```
Remove the entire `Attributes` section. They're not intended to be public attributes AFAIK.

jaywonchung replied:
For every class.

sharonsyh committed:
Update docs/measure/index.md

Add link to the push gateway

Co-authored-by: Jae-Won Chung <jwnchung@umich.edu>

jaywonchung commented on zeus/metric.py:
```diff
@@ -63,7 +85,7 @@ def __init__(
         job: str,
         gpu_bucket_range: Sequence[float] = [50.0, 100.0, 200.0, 500.0, 1000.0],
-        cpu_bucket_range: Sequence[float] = [10.0, 20.0, 50.0, 100.0, 200.0],
+        cpu_bucket_range: Sequence[float] = [10.0, 50.0, 100.0, 500.0, 1000.0],
```
You updated the default range, but you never reflected the change in any docstring or doc. (1) Update the docstring, and (2) remove the defaults listed in `measure.md` and instead point people to the API reference page for defaults.

sharonsyh committed:
Update docs/measure/index.md

Generalize the device as {component} with the note that Gauge only supports GPU

Co-authored-by: Jae-Won Chung <jwnchung@umich.edu>

sharonsyh committed:
Update docs/measure/index.md

Co-authored-by: Jae-Won Chung <jwnchung@umich.edu>

jaywonchung commented on zeus/metric.py:
```diff
@@ -109,111 +131,120 @@ def __init__(
-        self.energy_monitor.begin_window(
-            f"__EnergyHistogram_{name}", sync_execution=True
-        )
+        self.energy_monitor.begin_window(f"__EnergyHistogram_{name}", sync_execution)
```
```suggestion
        self.energy_monitor.begin_window(f"__EnergyHistogram_{name}", sync_execution=sync_execution)
```

jaywonchung replied:
Ditto for end_window.

sharonsyh committed:
Update docs/measure/index.md

Co-authored-by: Jae-Won Chung <jwnchung@umich.edu>

sharonsyh committed:
Update docs/measure/index.md

Co-authored-by: Jae-Won Chung <jwnchung@umich.edu>

sharonsyh committed:
Update docs/measure/index.md

Co-authored-by: Jae-Won Chung <jwnchung@umich.edu>

sharonsyh committed:
Update docs/measure/index.md

Co-authored-by: Jae-Won Chung <jwnchung@umich.edu>

jaywonchung commented on zeus/metric.py:
```diff
@@ -288,28 +319,36 @@ def begin_window(self, name: str) -> None:
                 self.update_period,
                 self.prometheus_url,
                 self.job,
+                sync_execution,
```
This is wrong. If `sync_execution` is `True`, you need to call `zeus.utils.framework.sync_execution` on the main thread there the application is running. On the other hand, the power/energy monitor process's `ZeusMonitor` should always be invoked with `sync_execution=False`.

Read `ZeusMonitor` to see how `sync_execution` (sometimes a boolean parameter and other times a function in `zeus.utils.framework`) is being used.

jaywonchung commented on zeus/metric.py:
```diff
@@ -288,28 +319,36 @@ def begin_window(self, name: str) -> None:
-    def end_window(self, name: str) -> None:
+        self.window_state[name] = MonitoringProcessState(
+            queue=self.queue, proc=self.proc
+        )
```
Why are you putting these in `self.queue` and `self.proc`??

jaywonchung commented on zeus/metric.py:
```diff
@@ -288,28 +319,36 @@ def begin_window(self, name: str) -> None:
+        if self.queue is not None:
+            self.queue.put("stop")
+        else:
+            raise RuntimeError("Queue is not initialized")
```
`self.queue` can the queue from any random window??? More specifically, it's going to be the queue that belongs to the most recently started window.

This level of quality is completely unacceptable. Please re-check the correctness of every line of code and documentation, and then ask for review.

jaywonchung reviewed (changes requested):
Left some comments on changes to make. I think they will be more or less straightforward ones. Let's hope this is the final round of change requests. Thanks!

sharonsyh committed:
Remove power_limit_optimizer and bring back the original code for image processing

sharonsyh committed:
Add sync execution function to handle synchronization during monitoring window operations

sharonsyh committed:
Changed variable name prometheus_url -> pushgateway_url

sharonsyh committed:
Adjust unit test functions to reflect changes in the sync_execution logic and multiprocessing window state management.

sharonsyh committed:
Set one of cpus' supportsGetDramEnergyConsumption to True

jaywonchung committed:
Merge branch 'master' into prometheus-metrics-update

jaywonchung committed:
Docstring, newline

jaywonchung committed:
Doc large refactor

jaywonchung reviewed (approved):
LGTM. Thanks for your work!
